src/caip-types.ts

```diff
@@ -0,0 +1,198 @@
+export type CaipChainId = `${string}:${string}`;
+
+export const CaipAccountIdStruct = pattern(string(), CAIP_ACCOUNT_ID_REGEX);
+export type CaipAccountId = `${string}:${string}:${string}`;
+
+export const CaipAccountIdArrayStruct = array(CaipAccountIdStruct);
+
+/**
+ * A chain descriptor.
+ */
+export const CaipChainStruct = object({
+  id: CaipChainIdStruct,
+  name: LimitedString,
+});
+export type CaipChain = {
+  id: CaipChainId;
+  name: string;
+};
+
+export const CaipNamespaceStruct = object({
+  /**
+   * A list of supported chains in the namespace.
+   */
+  chains: array(CaipChainStruct),
+
+  /**
+   * A list of supported RPC methods on the namespace, that a DApp can call.
+   */
+  methods: optional(array(LimitedString)),
+
+  /**
+   * A list of supported RPC events on the namespace, that a DApp can listen to.
+   */
+  events: optional(array(LimitedString)),
+});
+export type CaipNamespace = {
+  chains: CaipChain[];
+  methods?: string[];
+  events?: string[];
+};
+
+/**
+ * A CAIP-2 namespace, i.e., the first part of a chain ID.
+ */
+export const CaipNamespaceIdStruct = pattern(string(), CAIP_NAMESPACE_ID_REGEX);
+export type CaipNamespaceId = Infer<typeof CaipNamespaceIdStruct>;
+
+/**
+ * A CAIP-10 account address, i.e., the last part of the account ID.
+ */
+export const CaipAccountAddressStruct = pattern(
+  string(),
+  CAIP_ACCOUNT_ADDRESS_REGEX,
+);
+export type CaipAccountAddress = Infer<typeof CaipAccountAddressStruct>;
+
+/**
+ * Check if the given value is a CAIP-2 chain ID.
+ *
+ * @param value - The value to check.
+ * @returns Whether the value is a CAIP-2 chain ID.
+ */
+export function isCaipChainId(value: unknown): value is CaipChainId {
+  return is(value, CaipChainIdStruct);
+}
+
+/**
+ * Check if the given value is a CAIP-2 namespace ID.
+ *
+ * @param value - The value to check.
+ * @returns Whether the value is a CAIP-2 namespace ID.
+ */
+export function isCaipNamespaceId(value: unknown): value is CaipNamespaceId {
+  return is(value, CaipNamespaceIdStruct);
+}
+
+/**
+ * Check if a value is a {@link CaipNamespace}.
+ *
+ * @param value - The value to validate.
+ * @returns True if the value is a valid {@link CaipNamespace}.
+ */
+export function isCaipNamespace(value: unknown): value is CaipNamespace {
+  return is(value, CaipNamespaceStruct);
+}
+
+/**
+ * Check if the given value is a CAIP-10 account ID.
+ *
+ * @param value - The value to check.
+ * @returns Whether the value is a CAIP-10 account ID.
+ */
+export function isCaipAccountId(value: unknown): value is CaipAccountId {
+  return is(value, CaipAccountIdStruct);
+}
+
+/**
+ * Check if the given value is an array of CAIP-10 account IDs.
+ *
+ * @param value - The value to check.
+ * @returns Whether the value is an array of CAIP-10 account IDs.
+ */
+export function isCaipAccountIdArray(value: unknown): value is CaipAccountId[] {
+  return is(value, CaipAccountIdArrayStruct);
+}
```

Comment:
don't see this being used anywhere in snaps repo

https://github.com/search?q=repo%3AMetaMask%2Fsnaps+isCaipAccountIdArray&type=code

if not, maybe we get rid of it and the type

Reply:
Wait why? We should be using these soon enough I believe.

Reply:
This is checking if a value is **array** of Caip Account Ids. 

You could easily just `value.every(isCaipAccountId)` instead